recipes-ni/ni-base-system-image-tests/files/fs_permissions_known.py

```diff
@@ -0,0 +1,153 @@
+import argparse
+import csv
+import datetime
+import hashlib
+import mmap
+import os
+import re
+import subprocess
+import sys
+
+class Logger:
+    prefix_logs = []
+    logs = []
+    def log(self, log):
+        self.logs.append(log)
+
+    def prefix_log(self, log):
+        self.prefix_logs.append(log)
+
+    def report(self):
+        for log in self.prefix_logs:
+            print(log)
+        for log in self.logs:
+            print(log)
+
+def run_cmd(cmd):
+    output = subprocess.check_output(cmd)
+    return output.decode('utf-8')
+
+# Columns: Path, Mode, User, Group, Link Target
+fs_manifest_format = '%p\t%M\t%u\t%g\t%l'
+def get_fs_manifest():
+    search_dirs = ['/bin', '/boot', '/etc', '/lib', '/lib64', '/sbin', '/usr', '/var']
+    omit_dirs = ['/etc/natinst/niskyline/Data/Assets/Cache', '/lib/modules', '/var/cache', '/var/run', '/var/tmp', '/var/volatile']
+
+    omit_expr = []
+    for d in omit_dirs:
+        omit_expr += ['-path', d, '-o']
+    omit_expr.pop()
+
+    fs_manifest = run_cmd(
+        ['find']
+        + search_dirs
+        + ['-printf', fs_manifest_format + '\n']
+        + ['('] + omit_expr + [')', '-prune']
+    )
+    entries = {}
+    for entry in fs_manifest.splitlines():
+        entry_data = entry.split('\t')
+        entries[entry_data[0]] = {
+            'mode': entry_data[1],
+            'user': entry_data[2],
+            'group': entry_data[3],
+            'link': entry_data[4]
+        }
+    return entries
+
+def log_version_info(logger, label, codename, full_version, date, db_id):
+    logger.log(f'INFO: {label} = {codename} {full_version} from {date} with _id {db_id}')
+
+def check_known_manifests(manifest, logger):
+    logger.log('###### Checking expected file permissions against actual for known folders')
+    failed = False
+    with open('known-permissions.csv') as known_perm_file:
+        permissions = csv.reader(known_perm_file, delimiter = ',')
+        header_row = True
+        for row in permissions:
+            if header_row:
+                header_row = False
+                continue
+
+            # Allow executing shell script to get path name.
```

Comment:
Parsing this ourselves is a bit smelly and not very performant. Regexes would have problems with nesting, but could we find a Python library that permits nesting, or assume/enforce that there is no nesting so that we could use something like [`re.finditer`](https://docs.python.org/3/library/re.html?highlight=re#re.finditer)? Perhaps even something so crude as actually invoking the shell like `run_cmd(['bash', '-c', f'echo {path}'])` could work (with proper escaping), and also handle wildcards.

Reply:
Here's a solution:

```
def run_cmd(cmd):
    output = subprocess.check_output(cmd)
    return output.decode('utf-8')

def eval_path_expr(path):
    first_pass_cmd = [ 'bash', '-c', f'echo {path}' ]
    first_pass = run_cmd(first_pass_cmd).strip().split('\n')
    return glob.glob(first_pass)
```

The `bash -c` wasn't resolving wildcards, but the built-in glob library can do that for us